From a358a2d206d82995fdd0b52615baadc3d6512588 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jozef=20=C4=8C=C3=AD=C5=BE?= <jozef.ciz.ml@gmail.com>
Date: Mon, 13 Feb 2023 19:04:11 +0100
Subject: [PATCH] Better layout of breakpoint condition modal window

---
 TextEditor.cpp | 15 +++++++++------
 TextEditor.h   |  2 +-
 2 files changed, 10 insertions(+), 7 deletions(-)

diff --git a/TextEditor.cpp b/TextEditor.cpp
index cef93b9..bfff42d 100644
--- a/TextEditor.cpp
+++ b/TextEditor.cpp
@@ -3277,7 +3277,7 @@ void TextEditor::Render(const char* aTitle, const ImVec2& aSize, bool aBorder)
 				bool isEnabled = bkpt.mEnabled;
 				if (ImGui::Selectable("Condition")) {
 					mPopupCondition_Line = line;
-					mPopupCondition_Use = bkpt.mUseCondition;
+					mPopupCondition_Use = true;
 					memcpy(mPopupCondition_Condition, bkpt.mCondition.c_str(), bkpt.mCondition.size());
 					mPopupCondition_Condition[std::min<size_t>(511, bkpt.mCondition.size())] = 0;
 					openBkptConditionWindow = true;
@@ -3570,20 +3570,23 @@ void TextEditor::Render(const char* aTitle, const ImVec2& aSize, bool aBorder)
 
 	// breakpoint condition popup
 	if (openBkptConditionWindow)
-		ImGui::OpenPopup("Condition##condition");
+		ImGui::OpenPopup("Condition##condition_modal");
 
 	ImFont* font = ImGui::GetFont();
 	ImGui::PopFont();
 	ImGui::SetNextWindowSize(ImVec2(430, 175), ImGuiCond_Once);
-	if (ImGui::BeginPopupModal("Condition##condition")) {
+	if (ImGui::BeginPopupModal("Condition##condition_modal", NULL, ImGuiWindowFlags_AlwaysAutoResize)) {
 		ImGui::Checkbox("Use condition", &mPopupCondition_Use);
 
 		if (!mPopupCondition_Use) {
 			ImGui::PushItemFlag(ImGuiItemFlags_Disabled, true);
 			ImGui::PushStyleVar(ImGuiStyleVar_Alpha, ImGui::GetStyle().Alpha * 0.5f);
-		}
-		ImGui::InputText("Condition", mPopupCondition_Condition, 512);
-		if (!mPopupCondition_Use) {
+        }
+        ImGui::Text("Condition");
+        ImGui::InputTextMultiline(
+            "##condition_input", mPopupCondition_Condition, sizeof(mPopupCondition_Condition),
+            ImVec2(420, ImGui::GetTextLineHeightWithSpacing() * 6), ImGuiInputTextFlags_AllowTabInput);
+        if (!mPopupCondition_Use) {
 			ImGui::PopStyleVar();
 			ImGui::PopItemFlag();
 		}
diff --git a/TextEditor.h b/TextEditor.h
index 224358e..2beba51 100644
--- a/TextEditor.h
+++ b/TextEditor.h
@@ -738,7 +738,7 @@ private:
 
 	int mPopupCondition_Line;
 	bool mPopupCondition_Use;
-	char mPopupCondition_Condition[512];
+	char mPopupCondition_Condition[1024];
 
 	bool mCheckComments;
 	ErrorMarkers mErrorMarkers;
-- 
2.39.1

