From fb8f7b835e086fe8d875d064f62bd613c00669ce Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jozef=20=C4=8C=C3=AD=C5=BE?= <jozef.ciz.ml@gmail.com>
Date: Sun, 25 Feb 2024 23:47:00 +0100
Subject: [PATCH] Add replace to undo history

---
 TextEditor.cpp | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/TextEditor.cpp b/TextEditor.cpp
index 7474fec..efc55bb 100644
--- a/TextEditor.cpp
+++ b/TextEditor.cpp
@@ -3433,11 +3433,25 @@ void TextEditor::Render(const char* aTitle, const ImVec2& aSize, bool aBorder)
 				auto selStart = curPos, selEnd = curPos;
 				selEnd.mColumn += strlen(mFindWord);
 				SetSelection(curPos, selEnd);
+				SetCursorPosition(selEnd);
+
+				UndoRecord u;
+				u.mBefore = mState;
+
+				u.mRemoved = GetText(curPos, selEnd);
+				u.mRemovedStart = curPos;
+				u.mRemovedEnd = selEnd;
 				DeleteSelection();
+				
+				u.mAdded = mReplaceWord;
+				u.mAddedStart = curPos;
 				InsertText(mReplaceWord);
-				SetCursorPosition(selEnd);
+				u.mAddedEnd = GetActualCursorCoordinates();
 				mScrollToCursor = true;
 
+				u.mAfter = mState;
+				AddUndo(u);
+
 				ImGui::SetKeyboardFocusHere(0);
 			};
 
-- 
2.44.0

