From c0559a0d73b60ebc8df0ed6ae3eb513be95664bb Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jozef=20=C4=8C=C3=AD=C5=BE?= <jozef.ciz.ml@gmail.com>
Date: Wed, 20 Apr 2022 11:21:16 +0200
Subject: [PATCH] Quick fix for issue with multiline input text horizontal
 scrolling

---
 imgui_widgets.cpp | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/imgui_widgets.cpp b/imgui_widgets.cpp
index f84d12a3..7e6edab9 100644
--- a/imgui_widgets.cpp
+++ b/imgui_widgets.cpp
@@ -4658,7 +4658,9 @@ bool ImGui::InputTextEx(const char* label, const char* hint, char* buf, int buf_
 
             // Store text height (note that we haven't calculated text width at all, see GitHub issues #383, #1224)
             if (is_multiline)
-                text_size = ImVec2(inner_size.x, line_count * g.FontSize);
+                text_size = ImVec2(inner_size.x - style.FramePadding.x, line_count * g.FontSize);
+                // text_size = InputTextCalcTextSizeW(state->TextW.begin(), state->TextW.begin() + state->CurLenW);
+            // printf("%f %f %f %f\n", inner_size.x, inner_size.y, text_size.x, text_size.y);
         }
 
         // Scroll
@@ -4764,7 +4766,7 @@ bool ImGui::InputTextEx(const char* label, const char* hint, char* buf, int buf_
     {
         // Render text only (no selection, no cursor)
         if (is_multiline)
-            text_size = ImVec2(inner_size.x, InputTextCalcTextLenAndLineCount(buf_display, &buf_display_end) * g.FontSize); // We don't need width
+            text_size = ImVec2(inner_size.x - style.FramePadding.x, InputTextCalcTextLenAndLineCount(buf_display, &buf_display_end) * g.FontSize); // We don't need width
         else if (!is_displaying_hint && g.ActiveId == id)
             buf_display_end = buf_display + state->CurLenA;
         else if (!is_displaying_hint)
-- 
2.36.0

