From f2d031c7cce027c0a6873dbe20c9ad67995f5f73 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jozef=20=C4=8C=C3=AD=C5=BE?= <jozef.ciz.ml@gmail.com>
Date: Sun, 16 Apr 2023 14:39:02 +0200
Subject: [PATCH] Fix clipboard

---
 src/mode.cpp | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/mode.cpp b/src/mode.cpp
index 9fd732f6..b731979f 100644
--- a/src/mode.cpp
+++ b/src/mode.cpp
@@ -368,7 +368,7 @@ void ZepMode::AddKeyPress(uint32_t key, uint32_t modifierKeys)
     key &= 0xFF;
 
     // Keys in this range converted to UTF8.  I need to figure out how to generically receive UTF8 here, but this
-    // temporary fix enables £-sign and other specials to display and work correctly
+    // temporary fix enables ï¿½-sign and other specials to display and work correctly
     if (key >= 127 && key <= 255)
     {
 
@@ -1392,6 +1392,7 @@ bool ZepMode::GetCommand(CommandContext& context)
     }
     else if (mappedCommand == id_StandardPaste)
     {
+        context.pRegister = &GetEditor().GetRegister('+');
         if (context.currentMode == EditorMode::Visual)
         {
             auto range = GetInclusiveVisualRange();
@@ -1416,6 +1417,7 @@ bool ZepMode::GetCommand(CommandContext& context)
     }
     else if (mappedCommand == id_PasteAfter)
     {
+        context.pRegister = &GetEditor().GetRegister('+');
         if (!context.pRegister->text.empty())
         {
             // Already in visual mode, so replace the selection
@@ -1452,6 +1454,7 @@ bool ZepMode::GetCommand(CommandContext& context)
     }
     else if (mappedCommand == id_PasteBefore)
     {
+        context.pRegister = &GetEditor().GetRegister('+');
         if (!context.pRegister->text.empty())
         {
             // Already in visual mode, so replace the selection with whatever we copied
-- 
2.40.0

