From 2ad9e19fb8c090db7dc0bb059a9d9c10a445aeeb Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jozef=20=C4=8C=C3=AD=C5=BE?= <jozef.ciz.ml@gmail.com>
Date: Sun, 16 Apr 2023 14:39:02 +0200
Subject: [PATCH] Fix clipboard

---
 src/mode.cpp | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/mode.cpp b/src/mode.cpp
index 5093b9d0..873b8c5f 100644
--- a/src/mode.cpp
+++ b/src/mode.cpp
@@ -368,7 +368,7 @@ void ZepMode::AddKeyPress(uint32_t key, uint32_t modifierKeys)
     key &= 0xFF;
 
     // Keys in this range converted to UTF8.  I need to figure out how to generically receive UTF8 here, but this
-    // temporary fix enables £-sign and other specials to display and work correctly
+    // temporary fix enables ï¿½-sign and other specials to display and work correctly
     if (key >= 127 && key <= 255)
     {
 
@@ -1403,6 +1403,7 @@ bool ZepMode::GetCommand(CommandContext& context)
     else if (mappedCommand == id_StandardPaste)
     {
         GetEditor().ReadClipboard();
+        context.pRegister = &GetEditor().GetRegister('+');
 
         if (context.currentMode == EditorMode::Visual)
         {
@@ -1430,6 +1431,7 @@ bool ZepMode::GetCommand(CommandContext& context)
     else if (mappedCommand == id_PasteAfter)
     {
         GetEditor().ReadClipboard();
+        context.pRegister = &GetEditor().GetRegister('+');
 
         if (!context.pRegister->text.empty())
         {
@@ -1467,6 +1469,7 @@ bool ZepMode::GetCommand(CommandContext& context)
     }
     else if (mappedCommand == id_PasteBefore)
     {
+        context.pRegister = &GetEditor().GetRegister('+');
         if (!context.pRegister->text.empty())
         {
             // Already in visual mode, so replace the selection with whatever we copied
-- 
2.44.0

