From 13c3a69b501d2c6b4a76b47ca25c7c3b34997c1b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jozef=20=C4=8C=C3=AD=C5=BE?= <jozef.ciz.ml@gmail.com>
Date: Tue, 12 Apr 2022 22:57:05 +0200
Subject: [PATCH] Temporarily fix confusing behaviour when selecting all of
 empty document

---
 src/mode.cpp | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/mode.cpp b/src/mode.cpp
index 33c73298..1fbf5278 100644
--- a/src/mode.cpp
+++ b/src/mode.cpp
@@ -995,6 +995,8 @@ bool ZepMode::GetCommand(CommandContext& context)
         m_visualBegin = context.buffer.Begin();
         m_visualEnd = context.buffer.End();
         auto range = GetInclusiveVisualRange(false);
+        if (range.first == range.second)
+            context.commandResult.modeSwitch = DefaultMode();
         GetCurrentWindow()->GetBuffer().SetSelection(range);
         GetCurrentWindow()->SetBufferCursor(range.second);
         return true;
-- 
2.36.0

