From 947fa05db79d8d7c558fdf4ac1363e5e2c2728d8 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jozef=20=C4=8C=C3=AD=C5=BE?= <jozef.ciz.ml@gmail.com>
Date: Wed, 13 Apr 2022 09:46:26 +0200
Subject: [PATCH] Fix moving to end of line in insert mode

---
 src/mode.cpp | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/src/mode.cpp b/src/mode.cpp
index 1fbf5278..3bcc7ef9 100644
--- a/src/mode.cpp
+++ b/src/mode.cpp
@@ -728,6 +728,8 @@ bool ZepMode::GetCommand(CommandContext& context)
         return true;
     }
 
+    auto movementLineEndLocation = context.currentMode == EditorMode::Insert ? LineLocation::LineCRBegin : LineLocation::LineLastNonCR;
+
     if (mappedCommand == id_NormalMode)
     {
         // TODO: I think this should be a 'command' which would get replayed with dot;
@@ -945,19 +947,19 @@ bool ZepMode::GetCommand(CommandContext& context)
     }
     else if (mappedCommand == id_MotionDown)
     {
-        GetCurrentWindow()->MoveCursorY(context.keymap.TotalCount());
+        GetCurrentWindow()->MoveCursorY(context.keymap.TotalCount(), movementLineEndLocation);
         context.commandResult.flags |= CommandResultFlags::HandledCount;
         return true;
     }
     else if (mappedCommand == id_MotionUp)
     {
-        GetCurrentWindow()->MoveCursorY(-context.keymap.TotalCount());
+        GetCurrentWindow()->MoveCursorY(-context.keymap.TotalCount(), movementLineEndLocation);
         context.commandResult.flags |= CommandResultFlags::HandledCount;
         return true;
     }
     else if (mappedCommand == id_MotionRight)
     {
-        GetCurrentWindow()->SetBufferCursor(cursorItr.MoveClamped(context.keymap.TotalCount()));
+        GetCurrentWindow()->SetBufferCursor(cursorItr.MoveClamped(context.keymap.TotalCount(), movementLineEndLocation));
         context.commandResult.flags |= CommandResultFlags::HandledCount;
         return true;
     }
-- 
2.36.0

