From fd49a8401e4a758b4ab7975190e2529735f07f27 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jozef=20=C4=8C=C3=AD=C5=BE?= <jozef.ciz.ml@gmail.com>
Date: Thu, 14 Apr 2022 12:56:19 +0200
Subject: [PATCH] Fix syntax highlight of escaped strings (\\ not handled)

---
 src/syntax.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/syntax.cpp b/src/syntax.cpp
index 51a37f72..c4665246 100644
--- a/src/syntax.cpp
+++ b/src/syntax.cpp
@@ -281,7 +281,7 @@ void ZepSyntax::UpdateSyntax()
                     {
                         auto itrNext = itrString + 1;
                         // Ignore quoted
-                        if (*itrString == '\\' && *itrNext == ch)
+                        if (*itrString == '\\')
                         {
                             itrString++;
                         }
-- 
2.36.0

