From a9471e4a9a2fc93187fe8e9494c85d286ae5cd65 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jozef=20=C4=8C=C3=AD=C5=BE?= <jozef.ciz.ml@gmail.com>
Date: Tue, 19 Apr 2022 15:52:17 +0200
Subject: [PATCH] Increase width of line number panel

---
 include/zep/editor.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/zep/editor.h b/include/zep/editor.h
index d2b3913b..04fe166c 100644
--- a/include/zep/editor.h
+++ b/include/zep/editor.h
@@ -196,7 +196,7 @@ struct SyntaxProvider
 const float bottomBorder = 2.0f;
 const float textBorder = 2.0f;
 const float tabSpacing = 1.0f;
-const float leftBorderChars = 3;
+const float leftBorderChars = 4;
 
 #define DPI_VEC2(value) (value * GetEditor().GetDisplay().GetPixelScale())
 #define DPI_Y(value) (GetEditor().GetDisplay().GetPixelScale().y * value)
-- 
2.36.0

